<br><br>
<font color='FireBrick' size='7'><b> Double/Debiased Machine Learning</b></font>

# using IPUMS data

In [1]:
import pandas as pd

main_data_df = pd.read_feather("/Users/dadmehr/R/main_data.feather")

print(f"Number of rows: {main_data_df.shape[0]}")
print(f"Number of columns: {main_data_df.shape[1]}")

Number of rows: 43424806
Number of columns: 54


In [2]:
data = main_data_df[main_data_df['YEAR'] == 2021]
data = data[['INCTOT', 'SEX', 'EDUC', 'EDUCD', 'RACE', 'AGE','EMPSTAT', 'MARST']]

data.to_csv('IPUMS_2021_data.csv', index=False)

In [3]:
data = data.sample(frac=0.01)  # Sampling 1% of the dataset due to GitHub's 25MB file size limit for uploads.

In [4]:
data = data[data['INCTOT'] != 9999999] 

In [5]:
data

INCTOT  SEX  EDUC  EDUCD  RACE  AGE  EMPSTAT  MARST
41886913   25000.0    1     6     65     1   25        1      3
43184983  228200.0    2    10    101     7   59        1      1
41042551     700.0    1     4     40     1   16        3      6
40650240   60000.0    2    11    115     6   32        1      6
41206161   75000.0    1     6     63     1   23        1      6
...            ...  ...   ...    ...   ...  ...      ...    ...
42856477   31600.0    2     6     65     2   76        3      5
42281517     100.0    2     7     71     1   40        3      2
42513508    3000.0    2     6     63     1   68        3      5
43346425   28000.0    2     6     63     1   54        1      1
40233517   17540.0    2     8     81     1   67        3      5

[27451 rows x 8 columns]

In [6]:
#!pip install statsmodels econml

In [7]:
import statsmodels.api as sm

# Prepare the variables
X = data[['SEX']]  # Independent variables
y = data['INCTOT']  # Dependent variable

# Add a constant to the independent variables (for the intercept)
X = sm.add_constant(X)

# Fit the model
model = sm.OLS(y, X).fit()

# Print the summary of the regression model
print(model.summary())


                            OLS Regression Results                            
Dep. Variable:                 INCTOT   R-squared:                       0.023
Model:                            OLS   Adj. R-squared:                  0.023
Method:                 Least Squares   F-statistic:                     649.9
Date:                Tue, 23 Apr 2024   Prob (F-statistic):          1.04e-141
Time:                        15:47:45   Log-Likelihood:            -3.4435e+05
No. Observations:               27451   AIC:                         6.887e+05
Df Residuals:                   27449   BIC:                         6.887e+05
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       7.789e+04   1305.748     59.654      0.0

In [8]:
import statsmodels.api as sm

# Prepare the variables
X = data[['SEX', 'EDUCD', 'RACE', 'AGE','EMPSTAT', 'MARST']]  # Independent variables
y = data['INCTOT']  # Dependent variable

# Add a constant to the independent variables (for the intercept)
X = sm.add_constant(X)

# Fit the model
model = sm.OLS(y, X).fit()

# Print the summary of the regression model
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                 INCTOT   R-squared:                       0.238
Model:                            OLS   Adj. R-squared:                  0.238
Method:                 Least Squares   F-statistic:                     1427.
Date:                Tue, 23 Apr 2024   Prob (F-statistic):               0.00
Time:                        15:47:45   Log-Likelihood:            -3.4094e+05
No. Observations:               27451   AIC:                         6.819e+05
Df Residuals:                   27444   BIC:                         6.820e+05
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       3.655e+04   2163.296     16.895      0.0

In [9]:
import statsmodels.api as sm

# Prepare the variables
X = data[['SEX', 'AGE']]  # Independent variables
y = data['INCTOT']  # Dependent variable

# Add a constant to the independent variables (for the intercept)
X = sm.add_constant(X)

# Fit the model
model = sm.OLS(y, X).fit()

# Print the summary of the regression model
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                 INCTOT   R-squared:                       0.035
Model:                            OLS   Adj. R-squared:                  0.035
Method:                 Least Squares   F-statistic:                     493.7
Date:                Tue, 23 Apr 2024   Prob (F-statistic):          2.32e-211
Time:                        15:47:45   Log-Likelihood:            -3.4418e+05
No. Observations:               27451   AIC:                         6.884e+05
Df Residuals:                   27448   BIC:                         6.884e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       6.066e+04   1608.006     37.724      0.0

<font color='FireBrick' size='5'><b>Frisch–Waugh–Lovell (FWL) theorem</b></font>

In [10]:
import pandas as pd
import statsmodels.api as sm


# Define the variables
y = data['INCTOT']  # Income
x1 = data['SEX']  # Sex
x2 = data['AGE']  # Age

# Residualize y and x1 with respect to x2
y_resid = sm.OLS(y, x2).fit().resid
x1_resid = sm.OLS(x1, x2).fit().resid

# Regress y_resid on x1_resid
fwl_regression = sm.OLS(y_resid, x1_resid).fit()

# Print the FWL coefficient
print(fwl_regression.summary())

                                 OLS Regression Results                                
Dep. Variable:                      y   R-squared (uncentered):                   0.000
Model:                            OLS   Adj. R-squared (uncentered):              0.000
Method:                 Least Squares   F-statistic:                              3.901
Date:                Tue, 23 Apr 2024   Prob (F-statistic):                      0.0483
Time:                        15:47:45   Log-Likelihood:                     -3.4488e+05
No. Observations:               27451   AIC:                                  6.898e+05
Df Residuals:                   27450   BIC:                                  6.898e+05
Df Model:                           1                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

<font color='FireBrick' size='5'><b>Why different coefficients</b></font>

The Frisch-Waugh-Lovell (FWL) theorem holds under certain assumptions, and there are cases where these assumptions are not met, making the theorem inapplicable or invalid. Here are some scenarios where the FWL theorem does not hold:

- **Non-linear relationships**: The FWL theorem assumes a linear relationship between the variables. If the relationships are non-linear, the theorem does not apply.
- **Non-constant variance**: The theorem assumes constant variance of the error terms. If the variance is not constant, the FWL theorem is not applicable.
- **Correlated errors**: The FWL theorem assumes independent and identically distributed error terms. If the errors are correlated, the theorem does not hold.
- **Multicollinearity**: If the independent variables are highly correlated, the FWL theorem may not be applicable due to multicollinearity issues.
- **Omitted variable bias**: If there are omitted variables that are correlated with both the independent and dependent variables, the FWL theorem may not hold.
- **Measurement error**: If there are measurement errors in the independent variables, the FWL theorem may not be applicable.
- **Time-series data**: The FWL theorem is typically applied to cross-sectional data. If the data is time-series, additional considerations and modifications are necessary.

Generalized linear models: The FWL theorem is specific to linear regression models. If the relationship is modeled using a generalized linear model (e.g., logistic regression), **the theorem does NOT apply**.
It's essential to carefully evaluate the assumptions and limitations of the FWL theorem before applying it to your data.

# Using just AGE for control

In [11]:
import doubleml as dml
import pandas as pd
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split

# Load your data (replace 'your_data.csv' with your actual file name)

# Get the column names for the outcome, treatment, and control(s)
y_col = 'INCTOT'  # Outcome variable
d_col = 'SEX'  # Treatment variable
x_cols = ['AGE']  # Control variables

# Create the dataset for DoubleML
data_dml = dml.DoubleMLData(data, y_col=y_col, d_cols=[d_col], x_cols=x_cols)

# Initialize the machine learning algorithms for DML
ml_y = RandomForestRegressor()  # For the outcome variable
ml_d = RandomForestRegressor()  # For the treatment variable

# Initialize and fit the DoubleMLPLR model
dml_plr = dml.DoubleMLPLR(data_dml, ml_y, ml_d, n_folds=10)
dml_plr.fit()

# Retrieve the summary DataFrame
summary = dml_plr.summary

# Check the available column names
print(summary.columns)

# Extract the coefficient for the treatment variable (SEX)
sex_coef = summary['coef'].loc[d_col]

# Output the coefficient for SEX
print(f"The estimated coefficient for SEX is: {sex_coef}")

Index(['coef', 'std err', 't', 'P>|t|', '2.5 %', '97.5 %'], dtype='object')
The estimated coefficient for SEX is: -20940.18239343873


# Result

The analysis suggests that, on average, females tend to have a lower income than males by approximately 20,940.

# Using more than just AGE for control 

In [12]:
import doubleml as dml
import pandas as pd
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split

# Load your data (replace 'your_data.csv' with your actual file name)

# Get the column names for the outcome, treatment, and control(s)
y_col = 'INCTOT'  # Outcome variable
d_col = 'SEX'  # Treatment variable
x_cols = ['EDUCD', 'RACE', 'AGE', 'EMPSTAT', 'MARST']  # Control variables
#x_cols = ['AGE']  # Control variables

# Create the dataset for DoubleML
data_dml = dml.DoubleMLData(data, y_col=y_col, d_cols=[d_col], x_cols=x_cols)

# Initialize the machine learning algorithms for DML
ml_y = RandomForestRegressor()  # For the outcome variable
ml_d = RandomForestRegressor()  # For the treatment variable

# Initialize and fit the DoubleMLPLR model
dml_plr = dml.DoubleMLPLR(data_dml, ml_y, ml_d, n_folds=10)
dml_plr.fit()

# Retrieve the summary DataFrame
summary = dml_plr.summary

# Check the available column names
print(summary.columns)

# Extract the coefficient for the treatment variable (SEX)
sex_coef = summary['coef'].loc[d_col]

# Output the coefficient for SEX
print(f"The estimated coefficient for SEX is: {sex_coef}")

Index(['coef', 'std err', 't', 'P>|t|', '2.5 %', '97.5 %'], dtype='object')
The estimated coefficient for SEX is: -18726.349889742072


# Result

The estimated coefficient for SEX is approximately -18,726, which means that, on average, being female (SEX=2) is associated with a decrease of 18,726 in INCTOT compared to being male (SEX=1), holding all other variables constant.

In other words, the analysis suggests that, on average, females tend to have a lower income than males by approximately 18,726.

Note that the p-value is not shown in the output, but it can be inferred to be very small (likely less than 0.001) given the large t-statistic and the confidence interval. This suggests that the coefficient for SEX is statistically significant, meaning that the observed association between SEX and INCTOT is unlikely to be due to chance.